### PR TITLE
Piso uvm workspace

### PR DESCRIPTION
**Pull request by balbal1**

### CODE REVIEW

File: uvm/tx/PISO/driver_piso.sv [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+                seq_item_port.item_done();
+            end
+        endtask : run_phase
+
+        virtual task drive_item(sequence_item_piso rhs);
+            @(negedge vif.BitCLK);
+            //*************************//
+            // TODO: Drive Inputs Here //
+            // Test case 1: Load parallel data 10'b1010101010
+        vif.TxParallel_10 = 10'b1010101010;
```
Comment (balbal1) [inline uvm/tx/PISO/driver_piso.sv:37]: use the `rhs` sequence item to drive the inputs, you can use hardcoded values for edge cases but not for all the tests.

File: uvm/tx/PISO/monitor_piso.sv [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+            if (vif.Reset == 0) begin
+                // Reset: expected data should be 0
+                expected_data = 10'b0;
+                temp_reg = 10'b0;
+            end else begin
+                // Shift expected data by one position
+                expected_data = expected_data >> 1;
+                expected_data[9] = vif.TxParallel_10[9]; // shift in the first bit of parallel data
+                temp_reg = expected_data;
+                ap_serial.write(vif.Serial); // Send serial output for analysis
```
Comment (balbal1) [inline uvm/tx/PISO/monitor_piso.sv:54]: what is `ap_serial` ? also you should send the whole `resp` sequence item. this code will only send 1-bit by 1-bit to the scoreboard, you should sample 10-bits using a for loop and then send them together using `resp` variable.

File: uvm/tx/PISO/scoreboard_piso.sv [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+        virtual function void write_piso(sequence_item_piso packet);
+            piso_q.push_back(packet);
+            //**************************//
+            // TODO: Check Results Here ///**/
+            bit [9:0] expected_serial;
+            bit [9:0] actual_serial;
+
+            function new(string name = "uvm_piso_scoreboard", uvm_component parent = null);
+            super.new(name, parent);
```
Comment (balbal1) [inline uvm/tx/PISO/scoreboard_piso.sv:28]: don't define a function inside the `write_piso` task. you should use the received `packet` and check it's internal signals that you sampled at the monitor to check the correctness of the output

File: uvm/tx/PISO/sequence_item_piso.sv [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    import uvm_pkg::*;
+    `include "uvm_macros.svh"
+    import enums::*;
+    
+    class sequence_item_piso extends uvm_sequence_item ;
+        `uvm_object_utils (sequence_item_piso)
+
+        //***************************//
+        // TODO: Define Signals Here //
+        rand logic [9:0] TxParallel_10;
```
Comment (balbal1) [inline uvm/tx/PISO/sequence_item_piso.sv:11]: also define a signal for the output to store the serial output in it using the monitor

File: uvm/tx/PISO/assertions_piso.sv [hunk 5]
```diff
@@ -0,0 +1,32 @@
+module assertions_piso (piso_if.DUT _if);
+    
+    //*****************************//
+    // TODO: Write Assertions Here //
+    // Assertion for Reset Behavior
+assert property (@(negedge _if.Reset) disable iff (!_if.Reset) (_if.TxParallel_10 == 10'b0 && _if.bit_count == 4'b0))
+  else $fatal("Reset assertion failed: TxParallel_10 or bit_count not reset to zero.");
```
Comment (balbal1) [inline uvm/tx/PISO/assertions_piso.sv:7]: `bit_count` is not a signal in the interface

File: uvm/tx/PISO/assertions_piso.sv [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+// Assertion for Bit Counter Behavior
+assert property (@(posedge _if.BitCLK) disable iff (!_if.Reset)
+  (_if.bit_count >= 0 && _if.bit_count <= 9) // bit_count should be within the range 0-9
+  && (_if.bit_count == 0 || (_if.bit_count == $past(_if.bit_count) - 1)))
```
Comment (balbal1) [inline uvm/tx/PISO/assertions_piso.sv:11]: same as before, you can't use `bit_count` here. (same for the rest of assertions)

File: uvm/tcl_stacktrace.txt [hunk 7]
```diff
@@ -0,0 +1,26 @@
+Sat Nov 30 9:58:07 PM Egypt Standard Time 2024
```
Comment (balbal1) [inline uvm/tcl_stacktrace.txt:1]: delete this file

**Review by balbal1 — COMMENTED** [review]
the code isn't working, at gave me at least 60 errors please try running the code before committing.
run the script on questasim and fix all errors please.
also add add `UVM/` folder to .gitignore
